src/utils/PSMFeatureExtractor.cpp

```diff
@@ -0,0 +1,310 @@
+            <td VALIGN="middle" ROWSPAN=2> \f$ \longrightarrow \f$ PSMFeatureExtractor \f$ \longrightarrow \f$</td>
+            <td ALIGN = "center" BGCOLOR="#EBEBEB"> pot. successor tools </td>
+        </tr>
+        <tr>
+            <td VALIGN="middle" ALIGN = "center" ROWSPAN=1> @ref TOPP_PeptideIndexer</td>
+            <td VALIGN="middle" ALIGN = "center" ROWSPAN=1> @ref TOPP_PercolatorAdapter </td>
+        </tr>
+    </table>
+  </center>
+
+  <p>
+PSMFeatureExtractor is search engine sensitive, i.e. it's extra features
+vary, depending on the search engine. Thus, please make sure the input is
+compliant with TOPP SearchengineAdapter output. Also, PeptideIndexer compliant
+target/decoy annotation is mandatory.
+Currently supported search engines are Comet, X!Tandem, MSGF+.
+Mascot support is available but in beta development.
+  </p>
+
+  @note if you have extra features you want to pass to percolator, use the extra
+ flag and list the MetaData entries containing the extra features.
+
+  <B>The command line parameters of this tool are:</B>
+  @verbinclude UTILS_PSMFeatureExtractor.cli
+  <B>INI file documentation of this tool:</B>
+  @htmlinclude UTILS_PSMFeatureExtractor.html
+*/
+
+// We do not want this class to show up in the docu:
+/// @cond TOPPCLASSES
+
+class PSMFeatureExtractor :
+  public TOPPBase
+{
+public:
+  PSMFeatureExtractor() :
+    TOPPBase("PSMFeatureExtractor", "Computes extra features for each input PSM.", false)
+  {
+  }
+
+protected:
+  void registerOptionsAndFlags_()
+  {
+    registerInputFileList_("in", "<files>", StringList(), "Input file(s)", true);
+    setValidFormats_("in", ListUtils::create<String>("mzid,idXML"));
+    registerOutputFile_("out", "<file>", "", "Output file in mzid or idXML format", true);
+    setValidFormats_("out", ListUtils::create<String>("mzid,idXML"));
+    registerStringList_("extra", "<MetaData parameter>", vector<String>(), "List of the MetaData parameters to be included in a feature set for precolator.", false, false);
+    // setValidStrings_("extra", ?);
+    // TODO: add this MHC feature back in with TopPerc::hasMHCEnd_()
+    //registerFlag_("MHC", "Add a feature for MHC ligand properties to the specific PSM.", true);
+    registerFlag_("multiple_search_engines", "Combine PSMs from different search engines by merging on scan level.");
+    registerFlag_("skip_db_check", "Manual override to skip the check if same settings for multiple search engines were applied. Only valid together with -multiple_search_engines flag.", true);
+    registerFlag_("concat", "Naive merging of PSMs from different search engines: concatenate multiple search results instead of merging on scan level. Only valid together with -multiple_search_engines flag.", true);
+    registerFlag_("impute", "Will instead of discarding all PSM not unanimously detected by all SE, impute missing values by their respective scores min/max observed. Only valid together with -multiple_search_engines flag.", true);
```

Comment:
description of impute and limit_imputation is not clear. I think it would also make sense to introduce a topp subsection.